shared/src/main/scala/kinesis4cats/producer/Producer.scala

```diff
@@ -188,65 +176,90 @@ abstract class Producer[F[_], PutReq, PutRes](implicit
-          x.leftTraverse(F.raiseError)
-        } else {
-          logger
-            .warn(ctx.context)(
-              "All retries have been exhausted, and the final retry detected errors"
-            )
-            .as(x)
-        }
-      case x =>
-        logger
-          .debug(ctx.context)("Failures detected, retrying failed records")
-          .flatMap { _ =>
-            for {
-              _ <- F.sleep(retryDuration)
-              failedRecords <- F.fromOption(
-                x.left.flatMap { e =>
-                  e.errors.flatMap(errors => errors.right)
-                },
-                new RuntimeException(
-                  "Failed records empty, this should never happen"
-                )
+
+    for {
+      ref <- Ref.of(Producer.RetryState[PutRes](records, None))
+      finalRes <- retryingOnFailuresAndAllErrors(
+        config.retryPolicy,
+        (x: Producer.Res[PutRes]) =>
+          F.pure(x.isRight || x.left.exists(e => e.errors.exists(_.isLeft))),
+        (x: Producer.Res[PutRes], details: RetryDetails) =>
+          for {
+            failedRecords <- F.fromOption(
+              x.left.flatMap { e =>
+                e.errors.flatMap(errors => errors.right)
+              },
+              new RuntimeException(
+                "Failed records empty, this should never happen"
               )
-              res <- putWithRetry(
+            )
+            _ <- logger.debug(ctx.addEncoded("retryDetails", details).context)(
+              s"Failures with ${failedRecords.size} records detected, retrying failed records"
+            )
+            _ <- ref.update(current =>
+              Producer.RetryState(
                 failedRecords.map(_.record),
-                retries.map(_ - 1),
-                retryDuration
+                current.res.fold(Some(x))(y => Some(y.combine(x)))
+              )
+            )
+          } yield (),
+        (e: Throwable, details: RetryDetails) =>
+          logger.error(ctx.addEncoded("retryDetails", details).context, e)(
+            "Exception when putting records, retrying"
+          )
+      )(ref.get.flatMap(x => _put(x.inputRecords)))
+      _ <-
+        if (finalRes.left.exists(e => e.errors.exists(_.right.nonEmpty))) {
+          if (config.raiseOnExhaustedRetries)
+            finalRes.leftTraverse(F.raiseError[Unit]).void
+          else {
+            logger
+              .warn(ctx.context)(
+                "All retries have been exhausted, and the final retry detected errors"
               )
-            } yield x.combine(res)
           }
-    }
+        } else F.unit
+      _ <- ref.update(current =>
+        Producer.RetryState(
+          current.inputRecords,
+          current.res.fold(Some(finalRes))(x => Some(x.combine(finalRes)))
+        )
+      )
+      res <- ref.get.flatMap(x =>
+        F.fromOption(
+          x.res,
+          new RuntimeException("Result is empty, this should never happen")
+        )
+      )
```

Comment:
**suggestion:** `modify` instead of `update` + `get`? not only would it be more type safe, it'd also be fiber safe